### DISCUSSION TITLE
Query Builder not mapping results correctly with circular manytomany joins

### DISCUSSION THREAD

**Discussion by ThatsARapp** [mikro-orm/mikro-orm#5182] [General]
### Describe the bug

In the reproduction below you can see that the first test returns the incorrect post. The second test is working as expected but serves to illustrate how the first should respond.

### Reproduction

```ts
import { Collection, Entity, ManyToMany, PrimaryKey } from '@mikro-orm/core';
import { MikroORM } from '@mikro-orm/sqlite';

@Entity()
export class Author {
  @PrimaryKey()
  id!: number;

  @ManyToMany(() => Tag)
  tags = new Collection<Tag>(this);

  @ManyToMany(() => Post)
  posts = new Collection<Post>(this);
}

@Entity()
export class Tag {
  @PrimaryKey()
  id!: number;

  @ManyToMany(() => Post)
  posts = new Collection<Post>(this);

  @ManyToMany(() => Author, 'tags')
  authors = new Collection<Post>(this);
}

@Entity()
class Post {
  @PrimaryKey()
  id!: number;

  @ManyToMany(() => Tag, 'posts')
  tags = new Collection<Tag>(this);

  @ManyToMany(() => Author, 'posts')
  authors = new Collection<Author>(this);
}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    dbName: ':memory:',
    entities: [Author, Post, Tag],
  });
  await orm.schema.refreshDatabase();
});

afterAll(() => orm.close(true));

beforeEach(async () => {
  const fork = orm.em.fork();
  await orm.schema.clearDatabase();

  const author1 = fork.create(Author, {
    id: 1,
  });
  const author2 = fork.create(Author, {
    id: 2,
  });

  const tag1 = fork.create(Tag, { id: 1 });

  const post1 = fork.create(Post, { id: 1 });
  const post2 = fork.create(Post, { id: 2 });

  author1.tags.add(tag1);
  author1.posts.add(post1);
  tag1.posts.add(post1);

  author2.tags.add(tag1);
  author2.posts.add(post2);
  tag1.posts.add(post2);

  await fork.flush();
});

test('it should map the entities correctly', async () => {
  const fork = orm.em.fork();
  const result = await fork
    .qb(Author, 'a')
    .select('*')
    .leftJoinAndSelect('a.tags', 't')
    .leftJoinAndSelect('t.posts', 'p')
    .leftJoin('p.authors', 'a2')
    .where('a.id = a2.id')
    .orderBy({ 'a.id': 'DESC' })
    .getSingleResult();
  expect(result.tags[0].posts[0].id).toBe(2);
});

test('the above test should return the same result as this one', async () => {
  const fork = orm.em.fork();
  const result = await fork
    .qb(Author, 'a')
    .select('*')
    .leftJoinAndSelect('a.tags', 't')
    .leftJoinAndSelect('t.posts', 'p')
    .leftJoin('p.authors', 'a2')
    .where('a.id = a2.id')
    .orderBy({ 'a.id': 'DESC' })
    .execute();
  expect(fork.map(Author, result[0]).tags[0].posts[0].id).toBe(2);
});
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.1

### Node.js version

16

### Operating system

macOs

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
This is indeed a bit weird edge case, but after a closer observation, I am not entirely sure this is a bug. The problem is that you are partially loading the `posts` collection, and your query returns two different versions of the `posts` collection for `tag.id: 2`:

```ts
[
  {
    id: 2,
    tags: [
      { id: 1, posts: [ { id: 2 } ] }
    ]
  },
  {
    id: 1,
    tags: [
      { id: 1, posts: [ { id: 1 } ] }
    ]
  }
]
```

This works the same for both of your tests, the difference is that in the first one, you map the whole response to entities via `qb.getSingleResult()`. While the name might be a bit confusing, it does map over the full response and only returns the first row out of it. That means that it sees two rows for the `tag.id: 1` and merges its collections - this is how result mapping works internally to deal with cartesian products from to-many joins. In your second test case, you use `em.map` only on the first row, ignoring the second one, so this problem is not visible, if you iterate both rows, it will be there too.

```ts
  const result = await fork
    .qb(Author, 'a')
    .select('*')
    .leftJoinAndSelect('a.tags', 't')
    .leftJoinAndSelect('t.posts', 'p')
    .leftJoin('p.authors', 'a2')
    .where('a.id = a2.id') // this causes the partial loading
    .orderBy({ 'a.id': 'DESC' })
    .execute();
  const authors = result.map(r => fork.map(Author, r)); // map all results
  expect(authors[0].tags[0].posts[0].id).toBe(2); // fails now
```

I think this itself could be fixed by only mapping the first row when using `qb.getSingleResult()`, as with the v6 changes in this part we already have the cartesian product merge and only the first row is what needs to be considered, but the behavior described above is surely gonna stay.

**Reply by B4nan** [reply]
If you would want to deal with this and load multiple results, you'll have to create separate contexts for them, this would work fine:

```ts
const result = await fork
  .qb(Author, 'a')
  .select('*')
  .leftJoinAndSelect('a.tags', 't')
  .leftJoinAndSelect('t.posts', 'p')
  .leftJoin('p.authors', 'a2')
  .where('a.id = a2.id') // this causes the partial loading
  .orderBy({ 'a.id': 'DESC' })
  .execute();
const authors = result.map(r => em.fork().map(Author, r)); // each result has its own context
expect(authors[0].tags[0].posts[0].id).toBe(2); // works now
```

I fixed your original test case here: 0e56fe134775ed800bea99752a98ecd615c9d4a6

**Comment by ThatsARapp** [comment]
Amazing thank you that makes total sense.
